### DISCUSSION TITLE
.NET 6 - Wrong folder for runtime files after compile

### DISCUSSION THREAD

**Discussion by wh1sp3r** [cefsharp/CefSharp#3990] [General]
Hi!
I created a project for .NET 6 with Any CPU. Installed the latest version of CefSharp.WinForms.NETCore (97.1.61).

It adds a lot of files into the project (language packs (*.pak) and CefSharp.dll, Ijwhost.dll, libEGL.dll, libGLESv2.dll). I excluded them, it does nothing if there are present in the project or not.

When I run this project, it crashes silently and creates an event in the windows event log that CefSharp.dll is missing.
 
In debug folder after build, there is a runtimes folder, inside is win-x64. inside is lib and native. Lib contains netcoreapp3.1, if I copy the contents of this folder into the native folder... it works!

Project settings:
```
  <PropertyGroup>
    <OutputType>WinExe</OutputType>
    <TargetFramework>net6.0-windows</TargetFramework>
    <Nullable>enable</Nullable>
    <UseWindowsForms>true</UseWindowsForms>
    <ImplicitUsings>enable</ImplicitUsings>
    <PlatformTarget>AnyCPU</PlatformTarget>
  </PropertyGroup>
```

Thanks!

**Comment by amaitland** [comment]
> I created a project for .NET 6 with Any CPU. Installed the latest version of CefSharp.WinForms.NETCore (97.1.61).

Creating a winexe that targets AnyCpu is not recommend. AnyCpu exists for library projects in .Net 5/6. See https://github.com/cefsharp/CefSharp/issues/2796#issuecomment-772342933


Did you read the post installation instructions? https://github.com/cefsharp/CefSharp/blob/cefsharp/97/NuGet/PackageReference/Readme.txt#L10

There is a working example at https://github.com/cefsharp/CefSharp.MinimalExample you can modify the netcore project to target .net 6 for testing purposes. 

> It adds a lot of files into the project (language packs (*.pak) and CefSharp.dll, Ijwhost.dll, libEGL.dll, libGLESv2.dll). I excluded them, it does nothing if there are present in the project or not.

It doesn't actually add anything directly and these can likely be ignored. They should in fact be hidden, still waiting on Microsoft to provide a solution for that see https://github.com/NuGet/Home/issues/4856

If you set a RuntimeIdentifier this should go away.

**Reply by amaitland** [reply]
For background 

- https://github.com/dotnet/sdk/issues/1906
- https://github.com/dotnet/docs/issues/18371

**Reply by amaitland** [reply]
Commit https://github.com/cefsharp/CefSharp.MinimalExample/commit/7df3ceb6579ce043d164bdbc3267c7c5058969f8 adds conditional `TargetFramework` for `net6.0-windows`.

Event targeting `AnyCPU(not recommended)` and everything appears to be working as expected.

**Comment by wh1sp3r** [comment]
Hi, thank you for the information.

I did some changes to project settings.
Now it's:

``` 
<PropertyGroup>
    <OutputType>WinExe</OutputType>
    <TargetFramework>net6.0-windows</TargetFramework>
    <RuntimeIdentifier Condition="'$(RuntimeIdentifier)' == ''">$(NETCoreSdkRuntimeIdentifier)</RuntimeIdentifier>
    <Nullable>enable</Nullable>
    <UseWindowsForms>true</UseWindowsForms>
    <ImplicitUsings>enable</ImplicitUsings>
    <PlatformTarget>x64</PlatformTarget>
  </PropertyGroup>
```

When run, it crashes silently with different errors.
It crashes in libcef.dll without any specific error (info from event logger)

I FORGOT to mention, the form with the browser control is in a CLASS PROJECT and packed into the NUGET package. 
The above project settings are in the main project which uses this NUGET (using the same settings in class project).

 Thank you

**Reply by amaitland** [reply]
> I FORGOT to mention, the form with the browser control is in a CLASS PROJECT and packed into the NUGET package.

You'll need to provide more detail for me to know what's going on. 

> The above project settings are in the main project which uses this NUGET (using the same settings in class project).

Showing only a snippet isn't actually enough. Please include the package references

**Reply by wh1sp3r** [reply]
I will prepare minimal example.

Thanks for helping me sorting this out :-)

**Reply by amaitland** [reply]
Don't set a RuntimeIdentifier for your library project. See https://github.com/cefsharp/CefSharp.Wpf.HwndHost/blob/master/CefSharp.Wpf.HwndHost/CefSharp.Wpf.HwndHost.csproj#L7 for an example of a CefSharp library project (also supports. Net 4.6).

**Reply by amaitland** [reply]
> I will prepare minimal example.

No zip files as I won't open them.

**Reply by wh1sp3r** [reply]
Hi, I solved the problem.

I added only one project reference to the library:
`<PackageReference Include="CefSharp.WinForms.NETCore" Version="97.1.61" />`

this reference is using CefSharp.Common.NETCore, but without including it too, it does not work. So I also added

`<PackageReference Include="CefSharp.Common.NETCore" Version="97.1.61" />`

and it works!

**Reply by amaitland** [reply]
> this reference is using CefSharp.Common.NETCore,

The reference only includes `CefSharp.Common.NETCore` as it's a `HWND Host` based implementation for `WPF` (basically a different instance of `ChromiumWebBrowser`).
